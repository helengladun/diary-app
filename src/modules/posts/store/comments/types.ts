export enum PostCommentsTypes {
    GET_POST_COMMENTS = '@@posts/GET_POST_COMMENTS',
    GET_POST_COMMENTS_SUCCESS = '@@posts/GET_POST_COMMENTS_SUCCESS',
    GET_POST_COMMENTS_FAILURE = '@@posts/GET_POST_COMMENTS_FAILURE',

    CLEAR_POST_COMMENTS = '@@posts/CLEAR_POST_COMMENTS',
}