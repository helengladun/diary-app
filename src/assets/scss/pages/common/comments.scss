.comments {
  max-width: 444px;
  width: 100%;
  padding: 8px 19px 14px 18px;

  border: 1px solid $primaryBorderColor;
  background-color: $primaryLightColor;
  box-sizing: border-box;

  &__content {
    .comment:last-child {
      border-bottom: none;
    }
  }

  &__title {
    margin-bottom: 20px;
    text-align: left;
  }
}