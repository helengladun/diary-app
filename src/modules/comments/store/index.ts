import {combineReducers} from 'redux';
import {ICommentsState, reducer as commentsReducer} from "./all-comments";

export interface CommentsState {
    allComments: ICommentsState,
}

export const reducer = combineReducers<CommentsState>({
    allComments: commentsReducer,
});
