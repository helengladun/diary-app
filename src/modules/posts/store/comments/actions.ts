import {action} from "typesafe-actions";
import {PostCommentsTypes} from "./types";
import {IComment} from "../../../shared/models/IComment";

export const getPostComments = () => action(PostCommentsTypes.GET_POST_COMMENTS);
export const getPostCommentsSuccess = (data: IComment[]) => action(PostCommentsTypes.GET_POST_COMMENTS_SUCCESS, {data});
export const getPostCommentsFailure = (message: string) => action(PostCommentsTypes.GET_POST_COMMENTS_FAILURE, { message });

export const clearPostComments = () => action(PostCommentsTypes.CLEAR_POST_COMMENTS);
