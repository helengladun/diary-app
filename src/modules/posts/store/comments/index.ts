import { AnyAction, Reducer } from 'redux';
import { PostCommentsTypes } from './types';
import {merge} from 'ramda';
import {IComment} from "../../../shared/models/IComment";

export interface IPostCommentsState {
    entity: IComment[],
    pending: boolean,
    message: string,
}

const initialState: IPostCommentsState = {
    pending: false,
    entity: [],
    message: ''
};

export const reducer: Reducer<IPostCommentsState> = (state = initialState, action: AnyAction) => {
    switch (action.type) {
        case PostCommentsTypes.GET_POST_COMMENTS:
            return merge(state, {
                pending: true
            });
        case PostCommentsTypes.GET_POST_COMMENTS_SUCCESS:
            return merge(state, {
                pending: false,
                entity: action.payload.data
            });
        case PostCommentsTypes.GET_POST_COMMENTS_FAILURE:
            return merge(state, {
                pending: false,
                message: action.payload.message
            });
        case PostCommentsTypes.CLEAR_POST_COMMENTS:
            return merge(state, {
                pending: false,
                entity: []
            });
        default:
            return state;
    }
};