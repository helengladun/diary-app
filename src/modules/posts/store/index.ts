import {combineReducers} from 'redux';
import {IPostsState, reducer as postsReducer} from "./all-posts";
import {IPostCommentsState, reducer as postCommentsReducer} from "./comments";

export interface PostsState {
    allPosts: IPostsState,
    postComments: IPostCommentsState
}

export const reducer = combineReducers<PostsState>({
    allPosts: postsReducer,
    postComments: postCommentsReducer
});
