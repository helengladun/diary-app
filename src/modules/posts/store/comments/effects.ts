import {ThunkDispatch} from "redux-thunk";
import {IApplicationState} from "../../../store/rootReducer";
import {Action} from "redux";
import {getPostComments, getPostCommentsFailure, getPostCommentsSuccess } from "./actions";
import {PostService} from "../../../shared/services/PostService";

export const getPostCommentsEffect = (post_id: number) => (dispatch: ThunkDispatch<IApplicationState, void, Action>, getState: Function) => {
    dispatch(getPostComments());

    try {
        const res = PostService.getComments(post_id);
        dispatch(getPostCommentsSuccess(res));
    } catch (e) {
        const message = 'Error found. Please, try again.';
        dispatch(getPostCommentsFailure(message));
    }
};