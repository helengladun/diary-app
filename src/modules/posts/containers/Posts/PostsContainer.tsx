import React, {ChangeEvent, Component, FormEvent} from 'react';
import {IPost} from "../../../shared/models/IPost";
import {IComment} from "../../../shared/models/IComment";
import {isEmpty} from 'ramda';
import {connect} from "react-redux";
import {IApplicationState} from "../../../store/rootReducer";
import {addPostEffect, deletePostEffect, getPostsEffect} from "../../store/all-posts/effects";
import uuid from 'uuid';
import AddPostForm from "../../components/AddPostForm/AddPostForm";
import {Post} from "../../components/Post/Post";
import {addCommentEffect, getCommentsEffect} from "../../../comments/store/all-comments/effects";
import {getPostCommentsEffect} from "../../store/comments/effects";
import Comments from "../../../comments/containers/Comments/CommentsContainer";

interface PropsFromDispatch {
    getPostsEffect: Function,
    getCommentsEffect: Function,
    addPostEffect: Function,
    getPostCommentsEffect: Function,
    addCommentEffect: Function,
    deletePostEffect: Function
}

interface PropsFromState {
    posts: IPost[],
    comments: IComment[],
    postComments: IComment[]
}

interface IProps extends PropsFromDispatch, PropsFromState {
}

interface IState {
    isCommentsVisible: boolean,
    postText: string,
    checkedPostId: number
}

export class PostsContainer extends Component<IProps, IState> {
    constructor(props: IProps) {
        super(props);

        this.state = {
            isCommentsVisible: false,
            postText: '',
            checkedPostId: 0
        };


        this.getPostsList = this.getPostsList.bind(this);
        this.choosePostHandler = this.choosePostHandler.bind(this);
        this.changePostHandler = this.changePostHandler.bind(this);
        this.submitPostFormHandler = this.submitPostFormHandler.bind(this);
        this.removePostHandler = this.removePostHandler.bind(this);
    }

    componentDidMount(): void {
        this.props.getPostsEffect();
        this.props.getCommentsEffect();
    }

    choosePostHandler(post_id: number) {
        this.props.getPostCommentsEffect(post_id);
        this.setState(() => ({isCommentsVisible: true, checkedPostId: post_id}))
    }

    changePostHandler(e: ChangeEvent<HTMLInputElement>) {
        const text = e.target.value;
        this.setState(() => ({postText: text}));
    }

    submitPostFormHandler(e: FormEvent<HTMLFormElement>) {
        e.preventDefault();

        this.props.addPostEffect({
            id: uuid(),
            description: this.state.postText
        });

        this.setState(() => ({
            postText: ''
        }));
    }

    removePostHandler(post_id: number) {
        this.props.deletePostEffect(post_id);
        this.setState(() => ({isCommentsVisible: false}))
    }

    render() {
        const {posts, postComments} = this.props;

        return (
            <>
                <div className="posts">
                    <h2 className="title title_1">Items</h2>
                    <AddPostForm value={this.state.postText} submitHandler={this.submitPostFormHandler} changeHandler={this.changePostHandler}/>
                    {posts && !isEmpty(posts) && this.getPostsList(posts)}
                </div>
                {this.state.isCommentsVisible && <Comments data={postComments} postId={this.state.checkedPostId}/>}
            </>
        );
    }

    private getPostsList(items: IPost[]) {
        return (
            items.map((item: IPost) => {
                const comments = this.props.comments.filter((comment: IComment) => comment.post_id === item.id);
                return (<Post key={item.id} data={item} chooseHandler={this.choosePostHandler}
                              isChecked={item.id === this.state.checkedPostId}
                              comments_count={comments.length} removeHandler={this.removePostHandler}/>)
            })
        )
    }
}

const mapStateToProps = ({posts, comments}: IApplicationState) => ({
    posts: posts.allPosts.entity,
    comments: comments.allComments.entity,
    postComments: posts.postComments.entity
});

const mapDispatchToProps: PropsFromDispatch = {
    getPostsEffect,
    getCommentsEffect,
    getPostCommentsEffect,
    addPostEffect,
    addCommentEffect,
    deletePostEffect
};

export default connect(mapStateToProps, mapDispatchToProps)(PostsContainer);