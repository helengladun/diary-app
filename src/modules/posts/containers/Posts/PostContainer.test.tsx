import React from 'react';
import {shallow, mount, ShallowWrapper, ReactWrapper} from 'enzyme';
import {PostsContainer} from "./PostsContainer";
import {comments} from "../../../shared/fixtures/comments";
import {posts} from "../../../shared/fixtures/posts";

jest.mock('uuid', () => jest.fn(() => 234));

describe('PostContainer test', () => {

    let getPostsEffect: Function, getCommentsEffect: Function, getPostCommentsEffect: Function,
        addPostEffect: Function, addCommentEffect: Function, deletePostEffect: Function,
        wrapperShallow: ShallowWrapper<any, any, React.Component<{}, {}, any>>, wrapperMounted: ReactWrapper<any, any, React.Component<{}, {}, any>>;

    beforeEach(() => {
        getPostsEffect = jest.fn();
        getCommentsEffect = jest.fn();
        getPostCommentsEffect = jest.fn();
        addPostEffect = jest.fn();
        addCommentEffect = jest.fn();
        deletePostEffect = jest.fn();

        wrapperShallow = shallow(<PostsContainer
            getCommentsEffect={getCommentsEffect}
            getPostCommentsEffect={getPostCommentsEffect}
            addPostEffect={addPostEffect}
            addCommentEffect={addCommentEffect}
            deletePostEffect={deletePostEffect}
            getPostsEffect={getPostsEffect}
            posts={posts}
            comments={comments}
            postComments={[comments[0], comments[1]]}
        />);

        wrapperMounted = mount(<PostsContainer
            getCommentsEffect={getPostsEffect}
            getPostCommentsEffect={getPostCommentsEffect}
            addPostEffect={addPostEffect}
            addCommentEffect={addCommentEffect}
            deletePostEffect={deletePostEffect}
            getPostsEffect={getPostsEffect}
            posts={posts}
            comments={comments}
            postComments={[comments[0], comments[1]]}
        />);
    });

    test('should render PostsContainer correctly', () => {
        expect(wrapperShallow).toMatchSnapshot();
    });

    test('set getPostsEffect, getCommentsEffect to false  on componentDidMount', () => {
        expect(getCommentsEffect).toHaveBeenCalled();
        expect(getPostsEffect).toHaveBeenCalled();
    });

    test('fire AddPostForm onSubmit form ', () => {
        const value = 'text';

        wrapperMounted.find('input').simulate('change', {
            target: { value }
        });

        wrapperMounted.find('form').simulate('submit', {
            preventDefault: () => {}
        });

        expect(addPostEffect).toHaveBeenLastCalledWith({
            id: 234,
            description: value
        });
    });

    test('fire deletePostEffect on click remove button on post', () => {
        wrapperMounted.find('button.post__remove-btn').at(0).simulate('click');
        expect(deletePostEffect).toHaveBeenLastCalledWith(1);
    });

    // test('fire deletePostEffect onDelete item', () => {
    //     wrapperShallow.dive().find('.post').at(0).simulate('click');
    //     expect(getPostCommentsEffect).toHaveBeenLastCalledWith(1);
    // });
});