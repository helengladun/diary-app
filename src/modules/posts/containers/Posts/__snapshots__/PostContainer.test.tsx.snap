// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostContainer test should render PostsContainer correctly 1`] = `
<Fragment>
  <div
    className="posts"
  >
    <h2
      className="title title_1"
    >
      Items
    </h2>
    <AddPostForm
      changeHandler={[Function]}
      submitHandler={[Function]}
      value=""
    />
    <Post
      chooseHandler={[Function]}
      comments_count={1}
      data={
        Object {
          "description": "First item with custom name",
          "id": 1,
        }
      }
      isChecked={false}
      key="1"
      removeHandler={[Function]}
    />
    <Post
      chooseHandler={[Function]}
      comments_count={2}
      data={
        Object {
          "description": "Second item is active",
          "id": 2,
        }
      }
      isChecked={false}
      key="2"
      removeHandler={[Function]}
    />
  </div>
</Fragment>
`;
